numpy/core/function_base.py

```diff
@@ -121,58 +121,82 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
     if num < 0:
         raise ValueError("Number of samples, %s, must be non-negative." % num)
     div = (num - 1) if endpoint else num
-
-    # Convert float/complex array scalars to float, gh-3504
-    # and make sure one can use variables that have an __array_interface__, gh-6634
-    start = asanyarray(start) * 1.0
-    stop  = asanyarray(stop)  * 1.0
-
-    dt = result_type(start, stop, float(num))
-    if dtype is None:
-        dtype = dt
-
-    delta = stop - start
-    y = _nx.arange(0, num, dtype=dt).reshape((-1,) + (1,) * ndim(delta))
-    # In-place multiplication y *= delta/div is faster, but prevents the multiplicant
-    # from overriding what class is produced, and thus prevents, e.g. use of Quantities,
-    # see gh-7142. Hence, we multiply in place only for standard scalar types.
-    _mult_inplace = _nx.isscalar(delta)
-    if div > 0:
-        step = delta / div
-        if _nx.any(step == 0):
-            # Special handling for denormal numbers, gh-5437
-            y /= div
-            if _mult_inplace:
-                y *= delta
-            else:
-                y = y * delta
+    # make sure one can use variables that have an __array_interface__, gh-6634
+    start = asanyarray(start)
+    stop = asanyarray(stop)
+    # handle datetime64/timedelta64 as a special case
+    if start.dtype.kind in "mM":
+        if dtype is None:
+            dtype = result_type(start, stop)
+            dt = dtype
         else:
-            if _mult_inplace:
-                y *= step
+            dt = result_type(start, stop, dtype)
+        start = start.astype(dt)
+        stop = stop.astype(dt)
+        delta = stop - start
+        delta_dt = delta.dtype
+        delta = delta.astype(_nx.int64)
+        y = _nx.arange(0, num, dtype=_nx.int64).reshape(
+            (-1,) + (1,) * ndim(delta))
+        if div > 0:
+            step = delta / div
+            if _nx.any(step == _nx.array(0, asanyarray(step).dtype)):
```

Comment:
Why can't this be `step == 0` like in the regular code path?

Reply:
My version borrowed from the original PR #14700 which states:
```
        # The following check that step is not 0 must be done with
        # an explicit dtype at right hand side because comparing 
        # datetime or timedelta objects with (unitless) 0 is not
        # allowed
        if _nx.any(step == _nx.array(0, asanyarray(step).dtype)):
```

However, I find that changing this line to `step == 0` is acceptable.

Reply:
Thanks for the reference. Please don't borrow weird code in future without also borrowing the comment that describes why it is weird!